File: Problem106/Solution.go
package main

type TreeNode struct {
	Val   int
	Left  TreeNode
	Right TreeNode
}

func buildTree(inorder []int, postorder []int) *TreeNode {
	postOrderIdx := len(postorder) - 1

	if postOrderIdx == -1 {
		return nil
	}
	root := &TreeNode{Val: postorder[postOrderIdx], Left: nil, Right: nil}

	rootIdx := 0
	for inorder[rootIdx] != postorder[postOrderIdx] {
		rootIdx++
	}
	root.Left = buildTree(inorder[:rootIdx], postorder[:rootIdx])
	root.Right = buildTree(inorder[rootIdx+1:], postorder[rootIdx:postOrderIdx])
	return root
}

File: Problem106/Solution.java
/**
 * @author tgorthi
 * @since Jun 2020
 */

import Solution.TreeNode;

public class Solution
{
    private int postOrderIdx;

    public TreeNode buildTree(int[] inorder, int[] postorder)
    {
        postOrderIdx = postorder.length - 1;
        return getRoot(inorder, postorder, 0, inorder.length - 1);
    }

    private TreeNode getRoot(int[] inorder, int[] postorder, int start, int end)
    {
        if (start > end)
        {
            return null;
        }
        TreeNode root = new TreeNode(postorder[postOrderIdx--]);
        int rootIdx = 0;
        while (inorder[rootIdx] != root.val)
        {
            rootIdx++;
        }
        root.right = getRoot(inorder, postorder, rootIdx + 1, end);
        root.left = getRoot(inorder, postorder, start, rootIdx - 1);
        return root;
    }
}